### PR TITLE
refactor(poetry): More dependency parsing refactoring

### PR DESCRIPTION
**Pull request by zharinov**
## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [x] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by zharinov** [comment]
- Blocked by #23965

File: lib/modules/manager/poetry/schema.ts [hunk 1]
```diff
@@ -132,58 +134,151 @@ export const PoetryDependencyRecord = LooseRecord(
...
       }
+      deps.push(dep);
     }
-  }
-  return record;
-});
-
+    return deps;
+  })
+  .catch([]);
```
Comment (viceice) [inline lib/modules/manager/poetry/schema.ts:155]: why catch? what can go wrong? we should consider that whole file as invalid when we can't parse it

File: lib/modules/manager/poetry/schema.ts [hunk 2]
```diff
@@ -132,58 +134,151 @@ export const PoetryDependencyRecord = LooseRecord(
...
+    for (const [groupName, group] of Object.entries(record)) {
+      for (const dep of Object.values(group)) {
+        dep.depType = groupName;
+        deps.push(dep);
+      }
+    }
+    return deps;
   })
-);
+  .catch([]);
```
Comment (viceice) [inline lib/modules/manager/poetry/schema.ts:186]: same here. what can go wrong?

Comment (zharinov) [reply lib/modules/manager/poetry/schema.ts:183]: These schemas are being used in the optional fields and we're merging them all into the single array. Unlike objects, array spread operator doesn't work well with undefined values, so I think the better course of action is to fallback to `[]`.

File: lib/modules/manager/poetry/schema.ts [hunk 3]
```diff
@@ -132,58 +132,152 @@ export const PoetryDependencyRecord = LooseRecord(
...
+        }
       }
+      deps.push(dep);
     }
-  }
-  return record;
-});
+    return deps;
+  })
+  .catch([]);
```
Comment (viceice) [inline lib/modules/manager/poetry/schema.ts:153]: still the question why catch here? do we really need to be so defensive? I would suggest to log a debug log with failed file parsing

Comment (zharinov) [reply lib/modules/manager/poetry/schema.ts:153]: For example, if we have `dependencies` field, but `devDependencies` is absent, we don't want to fail the entire parsing process. This schema is being used for 3 fields that aren't guaranteed to be present. I don't think we need to log their absense.

Comment (zharinov) [reply lib/modules/manager/poetry/schema.ts]: Okay, I've changed it to something more clear

**Review by rarkins — DISMISSED** [review]

**Review by rarkins — APPROVED** [review]

**Review by rarkins — APPROVED** [review]

**Pull request merged**